File: common/src/main/java/net/skullian/skyfactions/common/util/SLogger.java
package net.skullian.skyfactions.common.util;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import net.skullian.skyfactions.common.api.SkyApi;

public class SLogger {

    private static final String PREFIX = "<gray>[<reset><gradient:#0083FF:#00FFC7><bold>SkyFactions</gradient><reset><gray>]<reset>";

    public static void setup(Object message, boolean fatal, Object... args) {
        String text = fatal ? "<#e73f38>✗ㅤㅤ" + format(message, args) + "<#4294ed>" : "<#4294ed>➤ㅤㅤ" + format(message, args);
        String formatted = getFormatted(text);

        SkyApi.getInstance().getConsoleAudience().sendMessage(MiniMessage.miniMessage().deserialize("<bold>" + formatted + "<reset>"));
    }

    public static String getFormatted(String text) {
        String without = PlainTextComponentSerializer.plainText().serialize(MiniMessage.miniMessage().deserialize(text));

        int totalLength = 62;
        String leftDelimiter = "│";
        String rightDelimiter = "│";
        int availableSpace = totalLength - (leftDelimiter.length() + rightDelimiter.length());

        if (without.length() > availableSpace) {
            without = without.substring(0, availableSpace);
        }

        int paddingLeft = (availableSpace - without.length()) / 2;
        int paddingRight = availableSpace - without.length() - paddingLeft;
        return leftDelimiter + getPadding(paddingLeft) + text + getPadding(paddingRight) + rightDelimiter;
    }

    public static void noPrefix(Object message, Object... args) {
        String formatted = format(message, args);
        SkyApi.getInstance().getConsoleAudience().sendMessage(MiniMessage.miniMessage().deserialize("<#4294ed>"+ "<bold>" + formatted + "<reset>"));
    }

    private static String getPadding(int count) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < count; i++) {
            sb.append(" ");
        }
        return sb.toString();
    }

    public static void info(Object message, Object... args) {
        Component infoLog = MiniMessage.miniMessage().deserialize(PREFIX + " <#4294ed>" + format(message, args) + "<#4294ed><reset>");
        SkyApi.getInstance().getConsoleAudience().sendMessage(infoLog);
    }

    public static void warn(Object message, Object... args) {
        Component warnLog = MiniMessage.miniMessage().deserialize(PREFIX + " <#f28f24>" + format(message, args) + "<#f28f24><reset>");
        SkyApi.getInstance().getConsoleAudience().sendMessage(warnLog);
    }

    public static void fatal(Object message, Object... args) {
        Component fatalLog = MiniMessage.miniMessage().deserialize(PREFIX + " <#e73f38>" + format(message, args) + "<#e73f38><reset>");
        SkyApi.getInstance().getConsoleAudience().sendMessage(fatalLog);
    }

    private static String format(Object message, Object... args) {
        if (message == null) {
            return null;
        }
        String formattedMessage = message.toString();
        for (Object arg : args) {
            if (arg == null) continue;
            formattedMessage = formattedMessage.replaceFirst("\\{}", arg.toString());
        }
        return formattedMessage;
    }
}

File: common/src/main/kotlin/net/skullian/skyfactions/common/util/SLogger.kt
package net.skullian.skyfactions.common.util

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import net.skullian.skyfactions.api.SkyApi.Companion.getInstance

/**
 * The logger class.
 * Yeah, you're welcome.
 */
object SLogger {

    private var PREFIX = "<gray>[<reset><gradient:#0083FF:#00FFC7><bold>SkyFactions</gradient><reset><gray>]<reset>"
    private var DEBUG_PREFIX = "<gray>[<reset><bold>SkyFactions|DEBUG<reset><gray>]<reset>"

    /**
     * Print a warn log to console.
     */
    @JvmStatic
    fun info(message: Any?, vararg args: Any?) {
        val infoLog: Component = MiniMessage.miniMessage()
            .deserialize(PREFIX + " <#4294ed>" + format(message, args) + "<reset>")

        getInstance().getConsoleAudience().sendMessage(infoLog)
    }

    /**
     * Print a warn log to console.
     */
    @JvmStatic
    fun warn(message: Any?, vararg args: Any?) {
        val warnLog = MiniMessage.miniMessage()
            .deserialize(PREFIX + " <#f28f24>" + format(message, args) + "<reset>")
        getInstance().getConsoleAudience().sendMessage(warnLog)
    }

    /**
     * Print a fatal (error) log to console.
     */
    @JvmStatic
    fun fatal(message: Any?, vararg args: Any?) {
        val fatalLog = MiniMessage.miniMessage()
            .deserialize(PREFIX + " <#e73f38>" + format(message, args) + "<reset>")
        getInstance().getConsoleAudience().sendMessage(fatalLog)
    }

    /**
     * Print a debug log to console.
     * This will only successfully send if the console option is enabled.
     */
    @JvmStatic
    fun debug(message: Any?, vararg args: Any?) {
        // todo debug check

        val debugLog = MiniMessage.miniMessage()
            .deserialize(DEBUG_PREFIX + " <gray>" + format(message, args) + "<reset>")
        getInstance().getConsoleAudience().sendMessage(debugLog)
    }

    private fun format(message: Any?, vararg args: Any): String? {
        if (message == null) {
            return null
        }

        var formattedMessage = message.toString()
        for (arg in args) {
            formattedMessage = formattedMessage.replaceFirst("\\{}".toRegex(), arg.toString())
        }
        return formattedMessage
    }
}